### PR TITLE
Fix #11140 - Added support for `fullBase` option in `::link()`.

### LINKED ISSUE #11140: HtmlHelper::link fullBase option
This is a (multiple allowed):

* [x] bug
* [x] enhancement
* [ ] feature-discussion (RFC)

* CakePHP Version: CakePHP Version: 3.4.13.
* Platform and Target: nginx/1.13.4 , MariaDB/10.1.26, php7.0-fpm/7.0.19-1

### What you did
Used `$options['fullBase' => true]` in `HtmlHelper::link()`.
Output linkdid not contain full base.
Method options are inconsistent as `fullBase` is accepted as an option in the rest of `HtmlHelper`/`UrlHelper` options.

### What happened
Output link did not contain full base.

### What you expected to happen
Output link contains full base.

### PR DESCRIPTION
**Pull request by huwcbjones**
Fixes #11140 
Includes test case for `fullBase`.

### CODE REVIEW

File: src/View/Helper/HtmlHelper.php [hunk 1]
```diff
@@ -346,7 +346,8 @@ public function link($title, $url = null, array $options = [])
     {
         $escapeTitle = true;
         if ($url !== null) {
-            $url = $this->Url->build($url);
+            $url = $this->Url->build($url, $options);
+            $options = array_diff_key($options, ['fullBase' => null]);
```
Comment (markstory) [inline src/View/Helper/HtmlHelper.php]: Could this be `unset($options['fullBase'])` as well?

Comment (huwcbjones) [reply src/View/Helper/HtmlHelper.php]: Yes, it could be - I looked at what was implemented in `::image()` and copied that.

Comment (huwcbjones) [reply src/View/Helper/HtmlHelper.php]: After quickly benchmarking it `unset` is about 80% quicker than `array_diff` in this case - didn't see that one coming :smirk:
Will change accordingly!

File: src/View/Helper/HtmlHelper.php [hunk 2]
```diff
@@ -346,7 +346,8 @@ public function link($title, $url = null, array $options = [])
     {
         $escapeTitle = true;
         if ($url !== null) {
-            $url = $this->Url->build($url);
+            $url = $this->Url->build($url, $options);
```
Comment (markstory) [inline src/View/Helper/HtmlHelper.php]: We should probably whitelist options going into the `build()` right now all the attributes for the link go in which will make it hard for us to add options to build() in the future.

Comment (huwcbjones) [reply src/View/Helper/HtmlHelper.php]: Is that something for the 4.x.x release, or should we start doing it now?
And if now, what options are going to be whitelisted?

Comment (markstory) [reply src/View/Helper/HtmlHelper.php]: I don't know when options will be added, if ever. As for a whitelist we could start with `fullBase`.

Comment (markstory) [reply src/View/Helper/HtmlHelper.php]: I changed my mind. The options for UrlHelper::build() have not changed in a very long time. We can deal with the changes later one.

**Pull request merged**